### ISSUE TITLE
Concurrency issue with ProtocolFactory

### ISSUE THREAD

**Issue by jnioche** [DigitalPebble/storm-crawler#89]
I've just found that when looking at the code with Pradan, [ProtocolFactory](https://github.com/DigitalPebble/storm-crawler/blob/master/core/src/main/java/com/digitalpebble/storm/crawler/protocol/ProtocolFactory.java) caches the Protocol instances so there is exactly one instance within a ProtocolFactory. 

In the FetcherBolt we have a single instance of ProtocolFactory, which is used by all the threads concurrently. The method  getProtocolOutput in the HTTP [protocol](https://github.com/DigitalPebble/storm-crawler/blob/master/core/src/main/java/com/digitalpebble/storm/crawler/protocol/http/HttpProtocol.java#L205) could be called concurrently. Any shared state like `ifModifiedSince` could then become corrupted.

Pradhan will write a test class to illustrate the issue. We could sync the method but that would create a bottleneck. A simpler approach would be for the ProtocolFactory to create a new instance of the protocol for every request, it might be inexpensive to do so.

**Comment by jnioche** [comment]
ifModifiedSince could be handled within the HttpResponse instead.

**Comment by jakekdodd** [comment]
Whoops, that's definitely an oversight on my part in implementing `ifModifiedSince`.

One option would be to overload the `HttpResponse` constructor, like:

```java
// If-Modified-Since not in metadata
HttpResponse response = new HttpResponse(this, u)

// If-Modified-Since in metadata
HttpResponse response = new HttpResponse(this, u, ifModifiedSince)
```

But I have a sneaking suspicion that, eventually, we'll have other metadata-dependent behaviors for HttpResponse (Gui's ETag example, for instance). So if we're going to overload the constructor, the third argument should probably be the `knownMetadata`; HttpResponse would then be responsible for checking the metadata for behavior-driving keys.

**Comment by jnioche** [comment]
+1 to passing the metadata. do you have time to fix this?

**Comment by jakekdodd** [comment]
Yep no problem, will be able to tackle it this evening

**Comment by pradhanbu** [comment]
Jake, I am writing a test case for this. If its fine with you I can sneak
in the fix along with that.

**Comment by jakekdodd** [comment]
Yes absolutely!

**Comment by jnioche** [comment]
Fixed by Pradhan in #90 
Thanks!
